dopamine/replay_memory/circular_replay_buffer.py

```diff
@@ -255,16 +283,64 @@ def add(self, observation, action, reward, terminal, *args):
         extra_storage_types.
     """
     self._check_add_types(observation, action, reward, terminal, *args)
+    trajectory_index = self._get_current_trajectory()
+    self._add_to_trajectory_buffer(
+        trajectory_index, observation, action, reward, terminal, *args)
+
+  def _add_to_trajectory_buffer(
+      self, trajectory_index, observation, action, reward, terminal, *args):
+    """Adds a transition to the trajectory buffer.
+
+    Transitions are added to a trajectory until a terminal step is encountered
+    or the trajectory size reaches `max_trajectory_buffer`, in which case the
+    trajectory stored to the buffer is added to the memory and emptied.
+
+    Args:
+      trajectory_index: int, index of the trajectory to add to.
+      observation: np.array with shape observation_shape.
+      action: int, the action in the transition.
+      reward: float, the reward received in the transition.
+      terminal: A uint8 acting as a boolean indicating whether the transition
+                was terminal (1) or not (0).
+      *args: All the elements in a transition.
+
+    Raises:
+      ValueError: If `transition_index` is not in the range
+        [0, len(self._trajectories)].
+    """
+    if not (0 <= trajectory_index < len(self._trajectories)):
+      raise ValueError(
+          '`trajectory_index` must be in the '
+          'range [0, {}[. Given {} instead.'.format(
+              len(self._trajectories), trajectory_index))
+    trajectory = self._trajectories[trajectory_index]
+    trajectory.append(
+        tuple([observation, action, reward, terminal] + list(args)))
+
+    self._trajectory_lengths[trajectory_index] += 1
+
+    if terminal or (self._max_trajectory_buffer and (
+        len(trajectory) >= self._max_trajectory_buffer)):
+      self._add_trajectory_to_memory(trajectory_index)
```

Comment:
Flushing the buffer as an occasional side-effect of adding an element to it makes me uncomfortable.  If I'm reading or debugging and want to know when experiences actually get written to the buffer, it's hard to track down.

But this is a problem.  /Somebody/ needs to flush the buffer.  Checking how others do it...

Python's builtin [`print` solves this](https://docs.python.org/3/library/functions.html#print) by relying on a passed (or implicit) filehandle to decide when to flush (when flush isn't forced by a flag).  This pattern doesn't seem useful.

Java's [`BufferedWriter` solves this](https://docs.oracle.com/javase/7/docs/api/java/io/BufferedWriter.html#write(char[],%20int,%20int)) by autoflushing on write per the size of the buffer as defined when it was created.  This is like what's done here (or would be, if a trajectory_buffer was an object).

:man_shrugging: 

I think, then, that I'm happy enough with autoflushing, and I'm happy with it as implemented.  I think I'd be slightly happier if *this* method was called `_add`, and the other methods were called...  I don't know. Maybe it's not even true I'd be happier.  Sometimes I'm a grouch.

Reply:
renamed to `_add`. And renamed initial `_add` to `_add_transition_to_memory`.